packages/langium/src/grammar/langium-grammar-validator.ts

```diff
@@ -276,14 +277,23 @@ export class LangiumGrammarValidator {
             const isDataType = isDataTypeRule(rule);
             const isInfers = !rule.returnType && !rule.dataType;
             const ruleTypeName = getTypeName(rule);
+
             if (!isDataType && ruleTypeName && types.has(ruleTypeName) === isInfers) {
-                const keywordNode = isInfers ? findNodeForKeyword(rule.$cstNode, 'infer') : findNodeForKeyword(rule.$cstNode, 'returns');
+
+                // whether the 'infer' keyword is explicitly present
+                const implicitInfers = rule.inferredType === undefined;
+
+                // produce relevant diagnostic for missing returns or invalid infers
                 accept('error', getMessage(ruleTypeName, isInfers), {
-                    node: rule.inferredType ?? rule,
-                    property: 'name',
-                    code: isInfers ? IssueCodes.InvalidInfers : IssueCodes.InvalidReturns,
-                    data: keywordNode && toDocumentSegment(keywordNode)
+                    node: rule,
+                    property: implicitInfers ? 'name' : 'inferredType',
+                    code: isInfers ? (implicitInfers ? IssueCodes.MissingReturns : IssueCodes.InvalidInfers) : IssueCodes.InvalidReturns,
+                    data: {
+                        suggestion: implicitInfers ? `${ruleTypeName} returns ${ruleTypeName}` : `returns ${ruleTypeName}`,
```

Comment:
This breaks for some reason, when the type is explicitly inferred:

![image](https://user-images.githubusercontent.com/4377073/202497914-a87fdbaf-3526-4bd4-b664-3a2a25b76860.png)

<details>
<summary>Repro</summary>

```
grammar Arithmetics

entry Module:
    'module' name=ID
    (statements+=Statement)*;

interface Evaluation {
    expression: Expression
}

Statement:
    Definition | Evaluation;

Definition:
    'def' name=ID ('(' args+=DeclaredParameter (',' args+=DeclaredParameter)* ')')?
    ':' expr=Expression ';';

DeclaredParameter:
    name=ID;

type AbstractDefinition = Definition | DeclaredParameter;

Evaluation infers Evaluation:
    expression=Expression ';';

Expression:
    Addition;

Addition infers Expression:
    Multiplication ({infer BinaryExpression.left=current} operator=('+' | '-') right=Multiplication)*;

Multiplication infers Expression:
    PrimaryExpression ({infer BinaryExpression.left=current} operator=('*' | '/') right=PrimaryExpression)*;

PrimaryExpression infers Expression:
    '(' Expression ')' |
    {infer NumberLiteral} value=NUMBER |
    {infer FunctionCall} func=[AbstractDefinition] ('(' args+=Expression (',' args+=Expression)* ')')?;

hidden terminal WS: /\s+/;
terminal ID: /[_a-zA-Z][\w_]*/;
terminal NUMBER returns number: /[0-9]+(\.[0-9]*)?/;

hidden terminal ML_COMMENT: /\/\*[\s\S]*?\*\//;
hidden terminal SL_COMMENT: /\/\/[^\n\r]*/;
```

</details>

Reply:
Hmm, I'll take another look, and add a test case to cover this.

Reply:
The `data` prop was not set correctly for the rule. It seems this PR feel far enough behind that the mechanics changed internally a bit.  I've rebased and rewritten it a bit to clean it up to help with this.